services/background-worker/src/Wire/BackendNotificationPusher.hs

```diff
@@ -61,7 +72,21 @@ pushNotification targetDomain (msg, envelope) = do
           <> [ skipChanThreadKilled,
                logRetries (const $ pure True) logErrr
              ]
-  recovering policy handlers $ const go
+  -- Ensure that the mvars are reset correctly.
+  UnliftIO.bracket_ (takeMVar mvar) (putMVar mvar ()) $ do
```

Comment:
Why reset them? Just use it once as a signal to stop retrying?

Reply:
These aren't used once, they are paired with the consuming callback, and will be used each time a message for that callback arrives. These MVars are also shared with the cleanup code so it can wait until a given in-flight message is finished processing before shutting down the AMQP connection.

Reply:
If we use them once there are several problems.
1) The next time the consumer is called it will block forever as the MVar is already empty/full. Nothing would be clearing it for the next invocation.
2) The cleanup thread would have a similar issue, it would know that the first invocation of the consumer is either running or finished, but it would have no idea about all invocations after that and would blindly close the connection to AMQP, potentially while notifications are being sent and before they can be ACKed correctly, leading to redelivery to services.
3) Memory leaks. If the MVar is used once per consumer invocation, then we will have track them somewhere until the cleanup code is called and it would have to check all of them to be sure that the consumers aren't running. Alternatively we'd need another thread to periodically clean out this MVar collection, but that just adds to the multithreading coordination problem we already have.